File: src/main/scala/crusader728/leetcode/twopointer/ShortestWordDistanceII244.scala
package scala.crusader728.leetcode.twopointer


object ShortestWordDistanceII244 {

  class WordDistance(_wordsDict: Array[String]) {
    val storage: Map[String, List[Int]] =
      _wordsDict.indices.foldRight(Map.empty[String, List[Int]]) {case (i, acc) => {
        val word = _wordsDict(i)
        acc + (word -> (i :: acc.getOrElse(word, List.empty)))
      }}

    def shortest(word1: String, word2: String): Int = {
      val s1 = storage.getOrElse(word1, List.empty)
      val s2 = storage.getOrElse(word2, List.empty)

      @scala.annotation.tailrec
      def go(current: Int, l1: List[Int], l2: List[Int]): Int = {
        if(l1.isEmpty || l2.isEmpty) {
          current
        } else {
          val h1 = l1.head
          val h2 = l2.head
          if(h1 >= h2) {
            go(Math.min(current, h1 - h2), l1, l2.tail)
          } else {
            go(Math.min(current, h2 - h1), l1.tail, l2)
          }
        }
      }

      go(Int.MaxValue, s1, s2)
    }

  }
}

File: src/main/withJava/crusader728/leetcode/twopointer/ShortestWordDistanceII244.java
package withJava.crusader728.leetcode.twopointer;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ShortestWordDistanceII244 {
    static class WordDistance {
        private Map<String, List<Integer>> wordIdxMap = new HashMap<>();

        public WordDistance(String[] wordsDict) {
            for(int i = 0; i < wordsDict.length; ++i) {
                wordIdxMap.putIfAbsent(wordsDict[i], new ArrayList<>());
                wordIdxMap.get(wordsDict[i]).add(i);
            }
        }
        
        public int shortest(String word1, String word2) {
            List<Integer> l1 = wordIdxMap.get(word1);
            List<Integer> l2 = wordIdxMap.get(word2);
            int i = 0;
            int j = 0;
            int min = Integer.MAX_VALUE;
            while(i < l1.size() && j < l2.size()) {
                int p1 = l1.get(i);
                int p2 = l2.get(j);
                if(p1 > p2) {
                    min = Math.min(p1 - p2, min);
                    j++;
                } else {
                    min = Math.min(p2 - p1, min);
                    i++;
                }
            }
            return min;
        }
    }
}
